File: easy/821.shortest-distance-to-a-character.py
#
# @lc app=leetcode id=821 lang=python3
#
# [821] Shortest Distance to a Character
#

# @lc code=start
class Solution:
    def shortestToChar(self, s: str, c: str) -> List[int]:
        res = []

        c_pos = -10000 #s.length <= 10^4
        i = 0
        for char in s:
            if char == c:
                c_pos = i
            res.append(i - c_pos)
            i = i + 1
        
        c_pos = 10000 #s.length <= 10^4
        i = i - 1
        for char in s[::-1]:
            if char == c:
                c_pos = i;
            res[i] = min(res[i], c_pos - i)
            i = i - 1
        
        return res
        
# @lc code=end

File: easy/821.shortest-distance-to-a-character.js
/*
 * @lc app=leetcode id=821 lang=javascript
 *
 * [821] Shortest Distance to a Character
 */

// @lc code=start
/**
 * @param {string} s
 * @param {character} c
 * @return {number[]}
 */
var shortestToChar = function(s, c) {
  let ans = new Array(s.length);
  let count = s.length;
  for (let i=0; i<s.length; i++) {
    count = s[i] === c ? 0 : count + 1;
    ans[i] = count;
  }

  count = s.length;
  for (let i=s.length-1; i>=0; i--) {
    count = s[i] === c ? 0 : count + 1;
    ans[i] = Math.min(ans[i], count);
  }

  //"abaa"\n"b"
  return ans;
};
// @lc code=end

